pkg/controller/deployment/deployment_controller.go

```diff
@@ -419,17 +427,45 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 	return fmt.Errorf("unexpected deployment strategy type: %s", d.Spec.Strategy.Type)
 }
 
-func (dc *DeploymentController) syncRecreateDeployment(deployment extensions.Deployment) error {
-	newRC, err := dc.getNewRC(deployment)
+// Rolling back to a revision; no-op if the toRevision is deployment's current revision
+func (dc *DeploymentController) rollback(deployment extensions.Deployment, toRevision int) error {
+	if toRevision < 0 {
+		return nil
+	}
+	newRC, allOldRCs, err := dc.getNewRCAndAllOldRCs(deployment)
 	if err != nil {
 		return err
 	}
+	allRCs := append(allOldRCs, newRC)
+	// If rollback revision is 0, rollback to the last revision
+	if toRevision == 0 {
+		toRevision = secondMaxRevision(allRCs)
+	}
+	for _, rc := range allRCs {
+		v, err := revision(rc)
+		if err != nil {
+			return fmt.Errorf("unable to extract revision from deployment's RCs: %v", err)
+		}
+		glog.V(4).Infof("Found rc with revision %d: %+v", v, rc)
+		if v == toRevision {
+			glog.V(4).Infof("Found a match!")
+			// rollback by copying podTemplate.Spec from the rc, and increment revision number by 1
+			// no-op if the the spec matches current deployment's podTemplate.Spec
+			if !api.Semantic.DeepEqual(deployment.Spec.Template.Spec, rc.Spec.Template.Spec) {
```

Comment:
I also noticed that you compare the pod specs here while in order to 1) match the intent of the deployment, we compare the podtemplate [spec](https://github.com/kubernetes/kubernetes/blob/63ec304e42c4f1bb6bd90d3f638cf1aafafe4064/pkg/util/deployment/deployment.go#L103) or 2) to compute the hash for the unique label we still use podtemplate [spec](https://github.com/kubernetes/kubernetes/blob/63ec304e42c4f1bb6bd90d3f638cf1aafafe4064/pkg/controller/deployment/deployment_controller.go#L546) (note that podtemplate spec has [objectMeta](https://github.com/kubernetes/kubernetes/blob/63ec304e42c4f1bb6bd90d3f638cf1aafafe4064/pkg/api/types.go#L1183)). Shouldn't we be consistent in all places? I don't feel strong on what is right but podtemplate spec seems like what you should be using here @bgrant0607